src/Classes/TradeQueryRequests.lua

```diff
@@ -0,0 +1,275 @@
+function TradeQueryRequestsClass:SendRequest(url, callback, params)
+	params = params or {}
+	local id = LaunchSubScript([[
+		local curl = require("lcurl.safe")
+		local easy = curl.easy()
+		local url, req_body, versionNumber, POESESSID = ...
+		local res_header = ""
+		local res_body = ""
+        local httpheader = {
+				'User-Agent: Path of Building/' .. versionNumber .. ' (contact: pob@mailbox.org)',
+				'Accept: application/json',
+				'Content-Type: application/json',
+			}
+        if POESESSID then
+            table.insert(httpheader, 'Cookie: POESESSID='..POESESSID)
+        end
+		if req_body then
+			easy:setopt{
+				postfields = req_body,
+				post = true,
+			}
+		end
+		easy:setopt{
+			url = url,
+			httpheader = httpheader,
+		}
+		easy:setopt_headerfunction(function(data)
+			res_header = res_header..data
+			return true
+		end)
+		easy:setopt_writefunction(function(data)
+			res_body = res_body..data
+			return true
+		end)
+		easy:perform()
+		easy:close()
+		return res_header, res_body
+	]], "", "", url, params.body, launch.versionNumber, params.poesessid)
+	if id then
+		launch:RegisterSubScript(id, function(header, body, errMsg)
+			callback({header = header, body = body}, errMsg)
+		end)
+	end
+end
+
+---Performs search and fetches results
+---@param league string
+---@param query string
+---@param callback fun(items:table, errMsg:string)
+---@param params table @ params = { callbackQueryId = fun(queryId:string) }
+function TradeQueryRequestsClass:SearchWithQuery(league, query, callback, params)
+	params = params or {}
+	self:PerformSearch(league, query, function(itemHashes, queryId, errMsg)
+		if errMsg then
+			return callback(nil, errMsg)
+		end
+		if params.callbackQueryId then
+			params.callbackQueryId(queryId)
+		end
+		self:FetchResults(itemHashes, queryId, callback)
+	end)
+end
+
+---Perform search and run callback function on returned item hashes.
+---Item info has to be fetched seperately 
+---@param league string
+---@param query string
+---@param callback fun(itemHashes:string[], queryId:string, errMsg:string)
+function TradeQueryRequestsClass:PerformSearch(league, query, callback)
+	table.insert(self.requestQueue["search"], {
+		url = "https://www.pathofexile.com/api/trade/search/"..league,
+		body = query,
+		callback = function(response, errMsg)
+			if errMsg then
+				return callback(nil, nil, errMsg)
+			end
+			local response = dkjson.decode(response)
+			if not response then
+				errMsg =  "Failed to Get Trade response"
+				return callback(nil, nil, errMsg)
+			end
+			if not response.result or #response.result == 0 then
+				if response.error then
+					if response.error.code == 2 then
+						errMsg = "Complex Query - Please provide your POESESSID"
+					elseif response.error.message then
+						errMsg = response.error.message
+					end
+				else
+					errMsg = "No Matching Results Found"
+				end
+				return callback(nil, nil, errMsg)
+			end
+			callback(response.result, response.id, errMsg)
```

Comment:
This feels weird that we're passing individual bits of the response back to the callback.  Why not just pass the whole response and let the implementing function take care of the pieces?

Reply:
/shrug

Reply:
Returning the full response just hands over the parsing responsibility to the caller. Initially, there were calls from UI code so I didn't want to leak that logic into UI (not the case anymore). Also seemed like a good idea to constrain functions as much as possible so the intent is clear. In this case, item hashes and the query id are the only required data to perform an item fetch.
Anyways, the change commit is shorter than this message if you prefer it that way. 😀